feat: add btc fee choice

alter-eggo opened the pull request:
> Try out this version of the Hiro Wallet - download [extension builds](https://github.com/hirosystems/wallet/actions/runs/4690047791).<!-- Sticky Header Marker -->

Add btc fee choice in send flow

alter-eggo commented on src/app/pages/send/send-crypto-asset-form/form/btc/btc-send-form.tsx:
```diff
@@ -78,6 +79,11 @@ export function BtcSendForm() {
               <HighFeeDrawer learnMoreUrl={HIGH_FEE_WARNING_LEARN_MORE_URL_BTC} />
               <Outlet />
+
+              {/* This is for testing purposes only, to make sure the form is ready to be submitted */}
```
@kyranjamie @fbwoolf  not sure how to do it better. in tests we need to wait until `calcMaxSpend` will provide valid result

fbwoolf replied:
Thinking on this, it doesn't work just to add a `waitFor` in the test? It looks like the sendMaxButton is only used in one test. Can you wait for a value in the amount input in the test itself?

alter-eggo replied:
>  Can you wait for a value in the amount input in the test itself
don't understand what you mean, how it will help here? 🤔

fbwoolf replied:
I think I confused this with the STX form not realizing what was happing with the `calcMaxSpend` for btc. Not sure I have a better solution than the one you are proposing. 🤔

fbwoolf commented on src/app/pages/send/send-crypto-asset-form/components/send-crypto-asset-form.layout.tsx:
```diff
@@ -8,7 +8,7 @@ export function SendCryptoAssetFormLayout({ children }: SendCryptoAssetFormLayou
     <Flex
       alignItems="center"
-      data-testId={SendCryptoAssetSelectors.SendForm}
+      data-testid={SendCryptoAssetSelectors.SendForm}
```
Does this need to be rebased? Didn't you already fix this?

alter-eggo replied:
yeah, already fixed

fbwoolf commented:
> 
> Add btc fee choice in send flow
>
Would you mind always including screenshots of the UI work you do in PRs?

fbwoolf commented on src/app/pages/send/send-crypto-asset-form/form/btc/btc-set-fee.tsx:
```diff
@@ -0,0 +1,41 @@
+  };
+}
+
+export function BtcSetFee() {
```
I think the naming throughout this PR needs to accommodate ordinal txs as well, so using `Bitcoin` in naming rather than `Btc` ...maybe to match nav `toChooseBitcoinFee` could be `BitcoinChooseFee` (and route to match)? Also, I think this needs to be located in the `Family` folder and reused for ordinal sends (see attached issue)? It might be that we need to move that form first to make sharing components b/w these forms work?

alter-eggo replied:
yeah, agree here

fbwoolf commented:
@alter-eggo this looks really good, but I think there is more work to add the fee selection to the ordinal flow too ...which, I think, might require refactoring/relocating that form code?

alter-eggo commented:
btc flow:
https://user-images.githubusercontent.com/46521087/228943117-5788fee5-5a62-4080-8b22-68021787ace5.mov

inscription flow (here I temporary removed validation locally to show flow):
https://user-images.githubusercontent.com/46521087/228943187-fb1e96d6-9cf7-4deb-b77b-7095aa170501.mov

markmhx commented:
Looking great!

markmhx commented:
Noticing this error message displays briefly upon transition?

<img width="522" alt="Screenshot 2023-03-31 at 18 53 07" src="https://user-images.githubusercontent.com/28991/229182412-786e263b-4c86-4e73-9f77-ff8ca77ef2c3.png">

alter-eggo commented:
@markmhx yeah, I did it purposefully to show the flow in the video, normally this error doesn't allow to continue the send flow

alter-eggo commented:
@fbwoolf @kyranjamie so we can merge this as is, implementing only btc set fees, and when https://github.com/hirosystems/wallet/issues/3399 is ready, I will add inscription fees choice in separate pr

kyranjamie commented on src/app/pages/send/send-crypto-asset-form/family/bitcoin/components/useBitcoinSetFees.ts:
```diff
@@ -0,0 +1,68 @@
+import { useMemo } from 'react';
```
rename file with snake casing

kyranjamie commented on src/app/pages/send/send-crypto-asset-form/components/fees-card.tsx:
```diff
@@ -0,0 +1,42 @@
+      transition={transition}
+      padding="extra-loose"
+      width="100%"
+      onMouseOver={() => setBgColor('#F9F9FA')}
```
Can we do this in css with `_hover={{ background: '#F9F9FA' }}`

kyranjamie commented on src/app/pages/send/send-crypto-asset-form/family/bitcoin/components/useBitcoinSetFees.ts:
```diff
@@ -0,0 +1,68 @@
+    if (!feeRate) return [];
+    const txSizer = new BtcSizeFeeEstimator();
+    const { txVBytes } = txSizer.calcTxSize({
+      input_count: uninscribedUtxos.length,
```
This logic assumes that the transaction is going to contain all uninscribed taproot UTXOs, which is likely not the case.

For calculating inscription transfer fees, we'll use TR utxos (though for now, just `1`).
For calculating btc transfers, the number of inputs is dependent on the amount being transferred. 

This is where it gets tricky. It could be that for a transfer of _x_ the low fee estimate, might use 3 UTXOs, and that a high fee estimate uses 5.

For this case, I think we'll need to run the coinselection in `determineUtxosForSpend` 3 times, with varying fee rates, to determine the selections.

kyranjamie reviewed (commented):
Great work Anastas, sorry for the slow review.

Happy to jump on a call to help chat through this implementation if you like.

kyranjamie reviewed (approved):
👌🏼

mica000 commented:
This is looking good @alter-eggo!

Wondering about the location of the footer. Is there a reason the bottom is bigger? 
<img width="408" alt="Screenshot 2023-04-11 at 11 40 53" src="https://user-images.githubusercontent.com/119523541/231121749-377c917b-59f0-4a8c-ae3b-36a583bd7063.png">



Ideally, this should work like this:
<img width="786" alt="Screenshot 2023-04-11 at 11 44 54" src="https://user-images.githubusercontent.com/119523541/231122089-a409619f-f03c-4bfe-aed9-15684a9922b3.png">

alter-eggo commented:
@mica000 not sure here, I'll fix

alter-eggo commented:
hm, there is hardcoded 106px height https://github.com/hirosystems/wallet/blob/58520fd2a432d1c1c91648eca2e9d4074108bbe1/src/app/pages/send/send-crypto-asset-form/components/form-footer.tsx#L20

@fbwoolf was there any specific reason to set it to 106?

mica000 commented:
@alter-eggo Are we not informing about this anymore?
"Fees are deducted from your balance, it won’t affect your sending amount."

<img width="554" alt="Screenshot 2023-04-11 at 13 24 31" src="https://user-images.githubusercontent.com/119523541/231147375-74666179-a6a1-4b83-98da-34e270ac6488.png">

mica000 commented:
<img width="413" alt="Screenshot 2023-04-11 at 11 41 42" src="https://user-images.githubusercontent.com/119523541/231147943-87664ba3-1460-4ec1-9e9a-4dffe5e1861e.png">
@alter-eggo Same applies to this as well.

alter-eggo commented:
@mica000 
> @alter-eggo Are we not informing about this anymore?

I added this today, you prob need to download new build

alter-eggo commented:
if we fix footer to bottom, user won't see the memo warning, until scroll down, which I think most of the users won't do, are we ok with it?

![Screenshot 2023-04-11 at 16 28 01](https://user-images.githubusercontent.com/46521087/231162270-ed77a17a-0580-4d8c-a9fa-b507ba011943.png)

kyranjamie commented on src/app/components/info-card/info-card.tsx:
```diff
@@ -68,35 +69,38 @@ export function InfoCardAssetValue({
-      borderRadius="10px"
-    >
-      {icon && <Box as={icon} size="32px" />}
+    <Box width="100%" px="loose">
```
Just a note that, if we ever want this component without the `px` it becomes as hassle. Better to add it separately on each use?

alter-eggo replied:
yeah, agree here, done

alter-eggo commented:
@mica000 I made all the ui fixes

mica000 commented:
@alter-eggo thanks! 

Spotted a small copy error: 
**Fees** are deducted from your balance, it won't affect your sending amount.

Currently, it seems the copy is left-aligned, could you also align this copy to center?

mica000 commented:
@alter-eggo the footer error seems to persist on the popover when sending BTC. For STX works fine tho.

<img width="418" alt="Screenshot 2023-04-12 at 12 47 02" src="https://user-images.githubusercontent.com/119523541/231435709-a5d484e8-365e-4ab0-8bcf-373ae04ed960.png">
<img width="415" alt="Screenshot 2023-04-12 at 12 47 20" src="https://user-images.githubusercontent.com/119523541/231435697-c1e1cb2d-bf5e-47c3-b90e-853d00ddcb72.png">

mica000 commented:
@alter-eggo when using the fullscreen view, the corners are gone:

<img width="531" alt="Screenshot 2023-04-12 at 12 45 14" src="https://user-images.githubusercontent.com/119523541/231436282-8d921bab-a24b-4205-b5fa-bb44fae7fe78.png">
<img width="497" alt="Screenshot 2023-04-12 at 12 45 31" src="https://user-images.githubusercontent.com/119523541/231436284-ad6af940-e75d-4fe3-980a-190898df6ed3.png">

alter-eggo commented:
@mica000 fixed

fbwoolf commented:
Is this one ready to merge/release yet?

alter-eggo commented:
I think yes, it's ready

fbwoolf commented:
> I think yes, it's ready

I'll let you merge it when ready.

mica000 commented:
@alter-eggo Almost done, but found this:
<img width="526" alt="Screenshot 2023-04-13 at 15 26 13" src="https://user-images.githubusercontent.com/119523541/231773698-3f05bcbb-723e-4709-a2a4-576be4c00efc.png">

alter-eggo commented:
@mica000 it happens bc I added padding-bottom to page as in popup mode content is hidden beside fixed bottom btn. 
fixed in full page mode, thanks a lot for all the reviews 🙌
